### PR TITLE
spring action with rotation and position target, rework grab script to use it

### PR DESCRIPTION
**Pull request by sethalves**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6cce8459847567f6508d4cbd0235710418b67ea2 available at: https://gist.github.com/7d5991593fc2faa54d42<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5088.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5088.dmg

**Comment by MaitreyaBuddha** [comment]
I notice you have merged into your work from highfidelity a few times. While this can prove functionally correct, it has the potential for confusion tracking down errors later. I always go for a linear history whenever properly possible, for readability and debugability. Linus Torvald would agree and elaborate: http://www.mail-archive.com/dri-devel@lists.sourceforge.net/msg39091.html

**Comment by sethalves** [comment]
@MaitreyaBuddha are you a robot?

File: libraries/entities/src/EntityActionInterface.cpp [hunk 1]
```diff
@@ -73,17 +80,68 @@ glm::vec3 EntityActionInterface::extractVec3Argument(QString objectName, QVarian
...
+    float w = wV.toFloat(&wOk);
+    if (!xOk || !yOk || !zOk || !wOk) {
+        qDebug() << objectName << "argument" << argumentName
+                 << "must be a map with keys of x, y, z, w and values of type float.";
+        ok = false;
+        return glm::quat();
+    }
+
+    return glm::quat(w, x, y, z);
+}
```
Comment (AndrewMeadows) [inline libraries/entities/src/EntityActionInterface.cpp]: extractQuatArgument() never set's **ok** to be **true**.  It only sets it **false** for the failure cases.

You could do something like
    ok = xOk && yOk && zOk && wOk;
and then pivot on the value of **ok** at line 127.

File: libraries/physics/src/ObjectActionSpring.cpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+                    rigidBody->setLinearVelocity(glmToBullet(glm::vec3()));
+                }
+            }
+
+            // handle rotation
+            if (_rotationalTargetSet) {
+                glm::quat bodyRotation = bulletToGLM(rigidBody->getOrientation());
+                // if qZero and qOne are too close to each other, we can get NaN for angle.
+                auto alignmentDot = glm::dot(bodyRotation, _rotationalTarget);
+                const float almostOne = 0.99999;
```
Comment (AndrewMeadows) [inline libraries/physics/src/ObjectActionSpring.cpp]: Make the literal numeric value a **float** by adding the **f** postfix.

File: libraries/physics/src/ObjectActionSpring.cpp [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                    glm::quat deltaQ = target * qZeroInverse;
+                    glm::vec3 axis = glm::axis(deltaQ);
+                    float angle = glm::angle(deltaQ);
+                    if (isNaN(angle)) {
+                        qDebug() << "ObjectActionSpring::updateAction angle =" << angle
+                                 << "body-rotation =" << bodyRotation.x << bodyRotation.y << bodyRotation.z << bodyRotation.w
+                                 << "target-rotation ="
+                                 << target.x << target.y << target.z<< target.w;
+                    }
+                    assert(!isNaN(angle));
```
Comment (AndrewMeadows) [inline libraries/physics/src/ObjectActionSpring.cpp]: Keep the assert but get rid of the debug log.  They both are detecting the same thing.

File: libraries/physics/src/ObjectActionSpring.cpp [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    if (rtOk) {
+        _rotationalTarget = rotationalTarget;
+        _rotationalTargetSet = true;
+
+        if (rscOk) {
+            _angularTimeScale = angularTimeScale;
+        } else {
+            _angularTimeScale = 0.1;
```
Comment (AndrewMeadows) [inline libraries/physics/src/ObjectActionSpring.cpp]: 0.1 --> 0.1f here and above.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 45c7cd4929681f09f3209a8a7f5dc074b0ed9b1f available at: https://gist.github.com/7d5991593fc2faa54d42<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5088.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5088.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4216308b9b230516df04c467fb192b7493303ffc available at: https://gist.github.com/7d5991593fc2faa54d42<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5088.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5088.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ba4fd7adf6b0b3238b2f82e043ddbf7eb49e5d30 available at: https://gist.github.com/7d5991593fc2faa54d42<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5088.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5088.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8bd80c511ee8df21053f7ab47c93105484317dca available at: https://gist.github.com/7d5991593fc2faa54d42<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5088.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5088.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8ed9a3ca02e94bb064a42cb216b9b5c863384cc2 available at: https://gist.github.com/7d5991593fc2faa54d42<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5088.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5088.dmg

**Pull request merged**